process/modules/cycling_infra.py

```diff
@@ -18,22 +26,221 @@ def __init__(self, cfg: Config):
+        for buffer_class, buffer_value in self._buffers.items():
+            buffered_items = target_infra_polys.loc[target_infra_polys["hierarkia_yksisuuntaisuus"].isin(self._buffer_class_yksisuuntaisuus_values[buffer_class])].copy()
+            buffered_items["geometry"] = buffered_items.buffer(buffer_value)
+            retval = gpd.GeoDataFrame(pd.concat([retval, buffered_items], ignore_index=True))
+
+        return retval
+
+    def _check_and_set_ylre_classes_id(self, lines: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+        attrs = ["ylre_class", "kadun_nimi"]
+        ylre_katualueet_dissolved = self._ylre_katualueet.dissolve(by=attrs)
+        joined_result = gpd.sjoin(lines, ylre_katualueet_dissolved, predicate='within')
+
+        retval = lines.merge(joined_result[['uuid', 'index_right0', 'index_right1']], how='left', left_on='uuid', right_on='uuid')
+
+        return retval
+
+    def _makeValid(self, data, geom_column):
+        data[geom_column] = data.apply(lambda row: make_valid(row[geom_column]) if not row[geom_column].is_valid else row[geom_column], axis=1)
+        return(data)
+
     def process(self):
         self._process_result_lines = self._lines
 
+        # Drop unnecessary data rows base on main and sub type
+        cycle_lines = self._drop_not_used_classes_base_on_main_and_sub_types(
+            self._droppable_types, self._process_result_lines
+        )
+
+        # Keep rows base on hierarchy list
+        hierarchy_lines = self._keep_rows_base_on_hierarchy_list(
+            self._hierarkia, self._process_result_lines
+        )
+        # Merge hierarchy and cycle lines
+        self._process_result_lines = pd.concat([hierarchy_lines, cycle_lines])
+
+        # Drop dublicates
+        self._process_result_lines.drop_duplicates(subset=["gml_id"], inplace=True)
+
+        # Mark objects which are within YLRE katualueet areas
+        self._process_result_lines = self._check_and_set_ylre_classes_id(self._process_result_lines)
+
         # Buffering configuration
         buffers = self._cfg.buffer(self._module)
         if len(buffers) != 1:
             raise ValueError("Unknown number of buffer values")
 
-        # buffer lines
+        # Buffer lines using buffer configuration
         target_infra_polys = self._process_result_lines.copy()
-        target_infra_polys["geometry"] = target_infra_polys.buffer(buffers[0])
+        target_infra_polys = self._buffering(target_infra_polys)
+
+        target_infra_polys.rename(columns={"index_right1": "kadun_nimi"}, inplace=True)
+        # Drop unnecessary columns
+        target_infra_polys = self._drop_unnecessary_columns(
+            self._dropped_columns, target_infra_polys
+        )
+
+        # Dissolve geometries base on attributes listed
+        attrs = ["paatyyppi", "alatyyppi", "silta_alikulku", "hierarkia", "yksisuuntaisuus", "kadun_nimi", ]
+        for attr in attrs:
+            target_infra_polys[attr] = target_infra_polys[attr].fillna("NULL")
```

Comment:
The `tormays_cycle_infra_polys` table contains "NULL" strings instead of null values. Perhaps because of this line?

```
haitaton=# select distinct yksisuuntaisuus from tormays_cycle_infra_polys ;
                yksisuuntaisuus
------------------------------------------------
 Yksisuuntainen digitointisuuntaan
 Kaksisuuntainen
 Yksisuuntainen (ei tietoa digitointisuunnasta)
 NULL
 Yksisuuntainen digitointisuuntaa vastaan
(5 rows)

haitaton=# update tormays_cycle_infra_polys set yksisuuntaisuus = NULL where yksisuuntaisuus='NULL';
UPDATE 786

haitaton=# select distinct yksisuuntaisuus from tormays_cycle_infra_polys ;
                yksisuuntaisuus
------------------------------------------------
 Yksisuuntainen (ei tietoa digitointisuunnasta)
 Yksisuuntainen digitointisuuntaa vastaan
 <null>
 Yksisuuntainen digitointisuuntaan
 Kaksisuuntainen
(5 rows)
```

Reply:
Fixed